File: c/src/orchestration.c
/*
 * orchestration.c — CircleAI.Orchestration (C11 port).
 *
 * LocalAgentDispatcher: a small per-role handler table. Dispatch routes to the
 * registered handler; a missing handler yields a Blocked SwarmResult with the
 * actionable message. RunQualityGate classifies "[CRITICAL]"/"[HIGH]"-prefixed
 * issues (case-insensitive) as blockers, the rest as warnings.
 *
 * Pure C11 + libc. No pthreads.
 */

#include "circle_ai/orchestration.h"
#include "board_common.h"

#include <stdio.h>

/* ── record free ────────────────────────────────────────────────────────── */

void ca_orch_task_free(ca_orch_task_t *t) {
    if (!t) return;
    free(t->id);
    free(t->description);
    cab_strv_free(t->input_keys, t->input_count);
    cab_strv_free(t->input_values, t->input_count);
    t->id = t->description = NULL;
    t->input_keys = t->input_values = NULL;
    t->input_count = 0;
}

void ca_orch_swarm_result_free(ca_orch_swarm_result_t *r) {
    if (!r) return;
    free(r->task_id);
    free(r->output);
    cab_strv_free(r->issues, r->issue_count);
    r->task_id = r->output = NULL;
    r->issues = NULL;
    r->issue_count = 0;
}

void ca_orch_quality_gate_free(ca_orch_quality_gate_t *g) {
    if (!g) return;
    cab_strv_free(g->blockers, g->blocker_count);
    cab_strv_free(g->warnings, g->warning_count);
    g->blockers = g->warnings = NULL;
    g->blocker_count = g->warning_count = 0;
}

/* ── AgentSwarmConfig ───────────────────────────────────────────────────── */

#define MINUTES_5_MS (5LL * 60 * 1000)

ca_orch_swarm_config_t ca_orch_swarm_config_default(void) {
    ca_orch_swarm_config_t c;
    c.max_concurrency = 4;
    c.task_timeout_ms = MINUTES_5_MS;
    c.require_review_pass_before_deploy = true;
    c.require_security_pass_before_deploy = true;
    return c;
}

/* DeviceTierDefaults.MaxConcurrency mapped onto the C device tiers. The C#
 * DeviceTier scheme is {Wearable, Phone, Tablet, Desktop, Workstation}; the C
 * ca_device_tier_t adds Laptop + Embedded. Laptop maps to Desktop's slot (8),
 * Embedded to the default (2), matching the numeric intent. */
static int max_concurrency_for(ca_device_tier_t tier, int cpu_cores) {
    switch (tier) {
        case CA_TIER_WEARABLE:    return 1;
        case CA_TIER_PHONE:       return 2;
        case CA_TIER_TABLET:      return 4;
        case CA_TIER_LAPTOP:      return 8;  /* ~ DeviceTier.Desktop */
        case CA_TIER_WORKSTATION: {
            int v = cpu_cores - 2;
            if (v < 1) v = 1;
            if (v > 16) v = 16;
            return v;
        }
        case CA_TIER_EMBEDDED:    return 2;  /* default branch */
    }
    return 2;
}

ca_orch_swarm_config_t ca_orch_swarm_config_for_device(ca_device_tier_t tier,
                                                       int cpu_cores) {
    ca_orch_swarm_config_t c = ca_orch_swarm_config_default();
    c.max_concurrency = max_concurrency_for(tier, cpu_cores);
    return c;
}

/* ── LocalAgentDispatcher ───────────────────────────────────────────────── */

typedef struct {
    bool               set;
    ca_orch_handler_fn handler;
    void              *ctx;
} handler_slot_t;

struct ca_orch_dispatcher {
    handler_slot_t handlers[4]; /* indexed by ca_orch_role_t */
    bool           disposed;
};

ca_orch_dispatcher_t *ca_orch_dispatcher_create(void) {
    return (ca_orch_dispatcher_t *)calloc(1, sizeof(ca_orch_dispatcher_t));
}
void ca_orch_dispatcher_destroy(ca_orch_dispatcher_t *d) {
    free(d);
}

int ca_orch_dispatcher_register(ca_orch_dispatcher_t *d, ca_orch_role_t role,
                                ca_orch_handler_fn handler, void *ctx) {
    if (!d || !handler) return -1;
    if ((int)role < 0 || (int)role > 3) return -1;
    d->handlers[role].set     = true;
    d->handlers[role].handler = handler;
    d->handlers[role].ctx     = ctx;
    return 0;
}

static const char *role_name(ca_orch_role_t r) {
    switch (r) {
        case CA_ORCH_ROLE_ENGINEERING: return "Engineering";
        case CA_ORCH_ROLE_OPERATIONS:  return "Operations";
        case CA_ORCH_ROLE_REVIEW:      return "Review";
        case CA_ORCH_ROLE_SECURITY:    return "Security";
    }
    return "Engineering";
}

/* Build the Blocked-no-handler SwarmResult. Returns -1 on OOM. */
static int blocked_result(ca_orch_swarm_result_t *out, const ca_orch_task_t *t) {
    memset(out, 0, sizeof(*out));
    out->role   = t->role;
    out->status = CA_ORCH_STATUS_BLOCKED;
    out->completed_at_ms = t->created_at_ms; /* deterministic: no UtcNow here */
    out->task_id = cab_strdup_empty(t->id);
    if (!out->task_id) return -1;

    char msg[256];
    snprintf(msg, sizeof(msg), "No handler registered for role %s.",
             role_name(t->role));
    out->output = cab_strdup(msg);
    if (!out->output) { ca_orch_swarm_result_free(out); return -1; }

    out->issues = (char **)calloc(1, sizeof(char *));
    if (!out->issues) { ca_orch_swarm_result_free(out); return -1; }
    char issue[256];
    snprintf(issue, sizeof(issue),
             "Register a handler for AgentRole.%s before dispatching.",
             role_name(t->role));
    out->issues[0] = cab_strdup(issue);
    if (!out->issues[0]) { ca_orch_swarm_result_free(out); return -1; }
    out->issue_count = 1;
    return 0;
}

int ca_orch_dispatcher_dispatch(ca_orch_dispatcher_t *d,
                                const ca_orch_task_t *task,
                                ca_orch_swarm_result_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!d || !task || !out) return -1;
    if (d->disposed) return -1; /* ObjectDisposedException */
    if ((int)task->role < 0 || (int)task->role > 3) return -1;

    handler_slot_t *slot = &d->handlers[task->role];
    if (slot->set)
        return slot->handler(slot->ctx, task, out);
    return blocked_result(out, task);
}

/* Case-insensitive StartsWith. */
static bool starts_with_ci(const char *s, const char *prefix) {
    if (!s || !prefix) return false;
    size_t pl = strlen(prefix);
    for (size_t i = 0; i < pl; ++i) {
        if (s[i] == '\0') return false;
        if (tolower((unsigned char)s[i]) != tolower((unsigned char)prefix[i]))
            return false;
    }
    return true;
}

int ca_orch_dispatcher_run_quality_gate(const ca_orch_swarm_result_t *result,
                                        ca_orch_quality_gate_t *out) {
    if (out) memset(out, 0, sizeof(*out));
    if (!result || !out) return -1;

    /* Two passes: blockers first (in order), then warnings (in order). */
    if (result->issue_count > 0) {
        out->blockers = (char **)calloc(result->issue_count, sizeof(char *));
        out->warnings = (char **)calloc(result->issue_count, sizeof(char *));
        if (!out->blockers || !out->warnings) {
            ca_orch_quality_gate_free(out);
            return -1;
        }
    }
    for (size_t i = 0; i < result->issue_count; ++i) {
        const char *iss = result->issues[i];
        bool is_blocker = starts_with_ci(iss, "[CRITICAL]") ||
                          starts_with_ci(iss, "[HIGH]");
        char **dst = is_blocker ? out->blockers : out->warnings;
        size_t *cnt = is_blocker ? &out->blocker_count : &out->warning_count;
        dst[*cnt] = cab_strdup_empty(iss);
        if (!dst[*cnt]) { ca_orch_quality_gate_free(out); return -1; }
        (*cnt)++;
    }
    out->passed = (out->blocker_count == 0);
    return 0;
}

void ca_orch_dispatcher_dispose(ca_orch_dispatcher_t *d) {
    if (d) d->disposed = true;
}

File: go/orchestration.go
// orchestration.go
//
// Ports CircleAI.Orchestration: the agent-swarm dispatch surface used by
// loki-mode.
//
//	AgentRole / AgentPriority / AgentStatus (enums) -> int consts (stable ordinals)
//	AgentTask / SwarmResult / QualityGateResult (records) -> value structs
//	AgentSwarmConfig (record + Default/ForDevice)   -> AgentSwarmConfig
//	IAgentDispatcher                                -> AgentDispatcher interface
//	LocalAgentDispatcher                            -> LocalAgentDispatcher
//	IncidentTrigger (static)                         -> package funcs
//
// LokiOrchestrator and SecurityOrchestrationBridge are the host-side scheduling
// layer built ON this dispatcher; only the dispatch contract + incident mapping
// are in the portable surface named by the work unit.
//
// AgentPriority ordinals are Critical=0, High=1, Normal=2, Low=3 (lower = more
// urgent), so "bump one rank" = decrement toward Critical, matching the C#.

package circleai

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

// AgentRole categorises an agent's domain responsibility. Ports AgentRole
// (declaration-order ordinals: Engineering=0, Operations=1, Review=2, Security=3).
type AgentRole int

const (
	// AgentRoleEngineering — writes, reviews, and fixes code.
	AgentRoleEngineering AgentRole = 0
	// AgentRoleOperations — infrastructure, deployments, incident response.
	AgentRoleOperations AgentRole = 1
	// AgentRoleReview — quality review, testing, acceptance criteria.
	AgentRoleReview AgentRole = 2
	// AgentRoleSecurity — security analysis and vulnerability assessment.
	AgentRoleSecurity AgentRole = 3
)

// String returns the C# enum member name (used in dispatch messages).
func (r AgentRole) String() string {
	switch r {
	case AgentRoleEngineering:
		return "Engineering"
	case AgentRoleOperations:
		return "Operations"
	case AgentRoleReview:
		return "Review"
	case AgentRoleSecurity:
		return "Security"
	default:
		return "Engineering"
	}
}

// AgentPriority is a task's execution urgency (lower value = higher urgency).
// Ports AgentPriority (explicit ordinals).
type AgentPriority int

const (
	// AgentPriorityCritical — blocks all other work.
	AgentPriorityCritical AgentPriority = 0
	// AgentPriorityHigh — current session.
	AgentPriorityHigh AgentPriority = 1
	// AgentPriorityNormal — arrival order.
	AgentPriorityNormal AgentPriority = 2
	// AgentPriorityLow — best-effort.
	AgentPriorityLow AgentPriority = 3
)

// AgentStatus is a task/result lifecycle status. Ports AgentStatus
// (declaration-order ordinals: Pending=0, Running=1, Passed=2, Failed=3, Blocked=4).
type AgentStatus int

const (
	// AgentStatusPending — created, not dispatched.
	AgentStatusPending AgentStatus = 0
	// AgentStatusRunning — executing.
	AgentStatusRunning AgentStatus = 1
	// AgentStatusPassed — completed, all gates passed.
	AgentStatusPassed AgentStatus = 2
	// AgentStatusFailed — completed with an error.
	AgentStatusFailed AgentStatus = 3
	// AgentStatusBlocked — halted by a gate or missing handler.
	AgentStatusBlocked AgentStatus = 4
)

// AgentTask is a single unit of work for a swarm. Ports the AgentTask record.
type AgentTask struct {
	ID          uuid.UUID
	Role        AgentRole
	Description string
	Priority    AgentPriority
	Inputs      map[string]string
	CreatedAt   time.Time
}

// NewAgentTask stamps a fresh AgentTask with a new UUID + current UTC time.
// Ports AgentTask.Create (nil inputs -> empty map).
func NewAgentTask(role AgentRole, description string, priority AgentPriority, inputs map[string]string) AgentTask {
	if inputs == nil {
		inputs = map[string]string{}
	}
	return AgentTask{
		ID:          uuid.New(),
		Role:        role,
		Description: description,
		Priority:    priority,
		Inputs:      inputs,
		CreatedAt:   time.Now().UTC(),
	}
}

// SwarmResult is an agent handler's outcome for a task. Ports the SwarmResult
// record.
type SwarmResult struct {
	TaskID      uuid.UUID
	Role        AgentRole
	Status      AgentStatus
	Output      string
	Issues      []string
	CompletedAt time.Time
}

// QualityGateResult is a gate verdict over a SwarmResult. Ports the
// QualityGateResult record.
type QualityGateResult struct {
	Passed   bool
	Blockers []string
	Warnings []string
}

// AgentSwarmConfig tunes swarm scheduling + gates. Ports the AgentSwarmConfig
// record.
type AgentSwarmConfig struct {
	MaxConcurrency                  int
	TaskTimeout                     time.Duration
	RequireReviewPassBeforeDeploy   bool
	RequireSecurityPassBeforeDeploy bool
}

// DefaultAgentSwarmConfig returns the production-safe defaults (4 concurrent,
// 5-minute timeout, both gates enforced). Ports AgentSwarmConfig.Default.
func DefaultAgentSwarmConfig() AgentSwarmConfig {
	return AgentSwarmConfig{
		MaxConcurrency:                  4,
		TaskTimeout:                     5 * time.Minute,
		RequireReviewPassBeforeDeploy:   true,
		RequireSecurityPassBeforeDeploy: true,
	}
}

// AgentSwarmConfigForDevice sizes MaxConcurrency by device tier, keeping the
// other defaults. Ports AgentSwarmConfig.ForDevice.
func AgentSwarmConfigForDevice(probe DeviceProbe) AgentSwarmConfig {
	return AgentSwarmConfig{
		MaxConcurrency:                  DeviceTierDefaults{}.MaxConcurrency(probe.Classify(), probe.CPUCores),
		TaskTimeout:                     5 * time.Minute,
		RequireReviewPassBeforeDeploy:   true,
		RequireSecurityPassBeforeDeploy: true,
	}
}

// AgentDispatcher routes tasks to handlers and runs quality gates. Ports
// IAgentDispatcher.
type AgentDispatcher interface {
	Dispatch(ctx context.Context, task AgentTask) (SwarmResult, error)
	RunQualityGate(ctx context.Context, result SwarmResult) (QualityGateResult, error)
}

// LocalAgentDispatcher is the in-process dispatcher routing tasks to handler
// funcs registered per AgentRole. Ports LocalAgentDispatcher. Construct with
// NewLocalAgentDispatcher; call Dispose to mark it disposed.
type LocalAgentDispatcher struct {
	mu       sync.Mutex
	handlers map[AgentRole]func(ctx context.Context, task AgentTask) (SwarmResult, error)
	disposed bool
}

// NewLocalAgentDispatcher constructs an empty dispatcher.
func NewLocalAgentDispatcher() *LocalAgentDispatcher {
	return &LocalAgentDispatcher{handlers: make(map[AgentRole]func(ctx context.Context, task AgentTask) (SwarmResult, error))}
}

// RegisterHandler registers (replacing) a handler for a role. Ports
// RegisterHandler. Panics if handler is nil.
func (d *LocalAgentDispatcher) RegisterHandler(role AgentRole, handler func(ctx context.Context, task AgentTask) (SwarmResult, error)) {
	if handler == nil {
		panic("handler must not be nil")
	}
	d.mu.Lock()
	d.handlers[role] = handler
	d.mu.Unlock()
}

// Dispatch routes a task to its role handler, or returns a Blocked result when
// no handler is registered. Ports DispatchAsync. Returns an error after Dispose.
func (d *LocalAgentDispatcher) Dispatch(ctx context.Context, task AgentTask) (SwarmResult, error) {
	d.mu.Lock()
	if d.disposed {
		d.mu.Unlock()
		return SwarmResult{}, errDispatcherDisposed
	}
	handler, ok := d.handlers[task.Role]
	d.mu.Unlock()
	if ok {
		return handler(ctx, task)
	}
	return SwarmResult{
		TaskID: task.ID,
		Role:   task.Role,
		Status: AgentStatusBlocked,
		Output: fmt.Sprintf("No handler registered for role %s.", task.Role),
		Issues: []string{fmt.Sprintf("Register a handler for AgentRole.%s before dispatching.", task.Role)},
		CompletedAt: time.Now().UTC(),
	}, nil
}

// RunQualityGate classifies [CRITICAL]/[HIGH]-prefixed issues (case-insensitive)
// as blockers and the rest as warnings. Ports RunQualityGateAsync.
func (d *LocalAgentDispatcher) RunQualityGate(ctx context.Context, result SwarmResult) (QualityGateResult, error) {
	blockers := make([]string, 0)
	warnings := make([]string, 0)
	for _, i := range result.Issues {
		if hasPrefixFold(i, "[CRITICAL]") || hasPrefixFold(i, "[HIGH]") {
			blockers = append(blockers, i)
		} else {
			warnings = append(warnings, i)
		}
	}
	return QualityGateResult{Passed: len(blockers) == 0, Blockers: blockers, Warnings: warnings}, nil
}

// Dispose marks the dispatcher disposed; subsequent Dispatch calls error. Ports
// Dispose (the C# completes an internal channel; the Go port has no queue to
// close, so Dispose only flips the disposed flag).
func (d *LocalAgentDispatcher) Dispose() {
	d.mu.Lock()
	d.disposed = true
	d.mu.Unlock()
}

// errDispatcherDisposed mirrors the C# ObjectDisposedException on a disposed
// LocalAgentDispatcher.
var errDispatcherDisposed = fmt.Errorf("LocalAgentDispatcher has been disposed")

// ── IncidentTrigger ─────────────────────────────────────────────────────────

var incidentCrashTags = toSet("crash", "exception", "unhandled_error", "oom", "null_reference")
var incidentSecurityTags = toSet("auth_failure", "permission_denied", "token_expired", "injection", "overflow")

// IncidentTasksFromMemoryEntry maps an episodic memory entry to the agent tasks
// that should fire for a crash / security incident. Ports
// IncidentTrigger.FromMemoryEntry — always one Operations task on a crash tag,
// plus one Security task when a security tag is also present; empty when not an
// incident.
func IncidentTasksFromMemoryEntry(entry EpisodicMemoryEntry) []AgentTask {
	tags := entry.Tags
	if tags == nil {
		tags = map[string]string{}
	}
	isCrash := false
	for k := range tags {
		if incidentCrashTags[strings.ToLower(k)] {
			isCrash = true
			break
		}
	}
	if !isCrash {
		return []AgentTask{}
	}
	appContext := ""
	if entry.AppContext != nil {
		appContext = *entry.AppContext
	}
	tasks := []AgentTask{
		NewAgentTask(
			AgentRoleOperations,
			"ops-incident: diagnose crash recorded at "+entry.RecordedAtUTC.Format(time.RFC3339Nano),
			AgentPriorityHigh,
			map[string]string{
				"episode_id":     entry.ID.String(),
				"user_text":      entry.UserText,
				"assistant_text": entry.AssistantText,
				"app_context":    appContext,
			}),
	}
	isSecurity := false
	for k := range tags {
		if incidentSecurityTags[strings.ToLower(k)] {
			isSecurity = true
			break
		}
	}
	if isSecurity {
		tasks = append(tasks, NewAgentTask(
			AgentRoleSecurity,
			"ops-security: investigate security incident from episode "+entry.ID.String(),
			AgentPriorityCritical,
			map[string]string{
				"episode_id":  entry.ID.String(),
				"app_context": appContext,
				"tags":        joinKeys(tags),
			}))
	}
	return tasks
}

// IncidentTaskFromAnomalySignal maps a confirmed anomaly signal to a security
// AgentTask, or returns (zero, false) when below dispatchThreshold. Ports
// IncidentTrigger.FromAnomalySignal — confidence drives priority, high-severity
// vectors bump one rank toward Critical.
func IncidentTaskFromAnomalySignal(signal *AnomalySignal, dispatchThreshold float64) (AgentTask, bool) {
	if signal == nil {
		panic("signal must not be nil")
	}
	if float64(signal.Confidence) < dispatchThreshold {
		return AgentTask{}, false
	}
	var priority AgentPriority
	switch {
	case signal.Confidence >= 0.85:
		priority = AgentPriorityCritical
	case signal.Confidence >= 0.60:
		priority = AgentPriorityHigh
	default:
		priority = AgentPriorityNormal
	}
	isHighSeverity := signal.Vector == ThreatVectorControlFlowDrift ||
		signal.Vector == ThreatVectorPrivilegeEscalation ||
		signal.Vector == ThreatVectorNetworkPivot ||
		signal.Vector == ThreatVectorStateCorruption
	if isHighSeverity && priority > AgentPriorityCritical {
		// Lower numeric = higher urgency; bump one rank toward Critical.
		p := int(priority) - 1
		if p < int(AgentPriorityCritical) {
			p = int(AgentPriorityCritical)
		}
		priority = AgentPriority(p)
	}
	inputs := make(map[string]string, len(signal.Evidence)+6)
	for k, v := range signal.Evidence {
		inputs[k] = v
	}
	inputs["signal_id"] = signal.ID
	inputs["vector"] = threatVectorName(signal.Vector)
	inputs["confidence"] = formatFloat3(float64(signal.Confidence))
	inputs["affected_module"] = signal.AffectedModule
	inputs["description"] = signal.Description
	inputs["detected_at"] = signal.DetectedAt.Format(time.RFC3339Nano)
	desc := fmt.Sprintf("ops-security: anomaly %s in %s (confidence %s)",
		threatVectorName(signal.Vector), signal.AffectedModule, formatPercent0(float64(signal.Confidence)))
	return NewAgentTask(AgentRoleSecurity, desc, priority, inputs), true
}

// joinKeys returns the map keys joined by "," (order is unspecified, mirroring
// the C# string.Join over a dictionary's keys).
func joinKeys(m map[string]string) string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	return strings.Join(keys, ",")
}

// formatFloat3 formats f with 3 fractional digits (invariant), matching the C#
// signal.Confidence.ToString("F3", InvariantCulture).
func formatFloat3(f float64) string {
	neg := f < 0
	if neg {
		f = -f
	}
	scaled := int64(f*1000 + 0.5)
	intPart := scaled / 1000
	frac := scaled % 1000
	d0 := byte('0' + (frac/100)%10)
	d1 := byte('0' + (frac/10)%10)
	d2 := byte('0' + frac%10)
	sign := ""
	if neg {
		sign = "-"
	}
	return sign + itoa64(intPart) + "." + string([]byte{d0, d1, d2})
}

// Interface guard.
var _ AgentDispatcher = (*LocalAgentDispatcher)(nil)
